cristin-commons/src/main/java/no/unit/nva/cristin/model/QueryBuilder.java

```diff
@@ -0,0 +1,236 @@
+    //            withPathIdentity(nonNullValue);
+    //        } else if (key.equals(PATH_ORGANISATION.getNvaKey())) {
+    //            withPathOrganization(nonNullValue);
+    //        } else if (key.equals(PATH_PROJECT.getNvaKey()) || key.equals(PATH_PROJECT.getKey())) {
+    //            withPathProject(nonNullValue);
+    //        } else {
+    //            invalidKeys.add(key);
+    //        }
+    //    }
+
+    /**
+     * Set Value by key, Key is validated.
+     */
+    protected abstract void setValue(String key, String value);
+    //    {
+    //        var qpKey = keyFromString(key,value);
+    //        if (!key.equals(qpKey.getKey()) && !query.isNvaQuery && qpKey != INVALID) {
+    //            query.isNvaQuery = true;
+    //        }
+    //        if(qpKey.equals(INVALID)) {
+    //            invalidKeys.add(key);
+    //        } else {
+    //            query.setValue(qpKey, value);
+    //        }
+    //    }
+
+    protected abstract Set<String> validKeys();
+    //    {
+    //        return query.isNvaQuery ? T.VALID_QUERY_PARAMETER_NVA_KEYS : T.VALID_QUERY_PARAMETER_KEYS;
+    //    }
+
+    protected void removeKey(T key) {
+        query.removeValue(key);
+    }
+
+    protected boolean invalidQueryParameter(T key, String value) {
+        return isNull(value) || !value.matches(key.getPattern());
+    }
+
+    protected boolean invalidPathParameter(T key, String value) {
+        return !(value.isBlank() || value.matches(key.getPattern()));
+    }
+
+    protected Set<String> getMissingKeys() {
+        return
+            requiredMissing()
+                .stream()
+                .map(qpk -> query.useNvaQuery ? qpk.getNvaKey() : qpk.getKey())
+                .collect(Collectors.toSet());
+    }
+
+    protected Set<T> required() {
+        return
+            Stream.concat(
+                    query.otherRequiredKeys.stream(),
+                    query.pathParameters.keySet().stream())
+                .collect(Collectors.toSet());
+    }
+
+    protected Set<T> requiredMissing() {
+        return
+            required().stream()
+                .filter(key -> !query.queryParameters.containsKey(key))
+                .filter(key -> !query.pathParameters.containsKey(key))
+                .collect(Collectors.toSet());
+    }
+
+    protected void throwInvalidParamererValue(Entry<T, String> entry) throws BadRequestException {
+        final var key = entry.getKey();
+        if (invalidQueryParameter(key, entry.getValue())) {
+            final var keyName = query.useNvaQuery ? key.getNvaKey() : key.getKey();
+            String errorMessage;
+            if (nonNull(key.getErrorMessage())) {
+                errorMessage = String.format(key.getErrorMessage(), keyName);
+            } else {
+                errorMessage = invalidQueryParametersMessage(keyName, EMPTY_STRING);
+            }
+            throw new BadRequestException(errorMessage);
```

Comment:
this gave me invalid title when removing language. I dont know why.

Reply:
removed language from project query, works now